metrics/sample.go

```diff
@@ -179,7 +180,7 @@ func parsePercentile(stat string) (float64, error) {
 // All methods should not panic, even if they are called on a nil pointer.
 //easyjson:skip
 type SampleTags struct {
-	tags map[string]string
+	tags *atlas.Node
 	json []byte
 }
```

Comment:
This enforces an additional allocation, for the SampleTags, object even if `tags` is "reused". 

There are currently two uses for this type:
1. caching `json`
2. Helps with the creation so that you cna use `AddTag` on the same object.

Arguably some encapsulation of `atlas.Node` although that likely isn't so intersting.

I would argue `2` is less of an issue and it can be solved with a type that can be allocated on the stack and then to retunr the `atlas.Node` that is then send away. How well that will work is to be seen - the k6 escape analysis might not work great on this case.

I would argue hte `json` caching likely is ... uh at best overrated. Most tags are on single samples either way so the cache won't be hit. (The cloud output outputs all 8 http metric sampels with 1 tag). 

This also totally doesn't take advantage of the fact that the vast majority of hte `atlas.Node` will be the same and consequently the `json` it produce will be as well.

Given That I kind of prefer if we drop that and move it to a `map[*atlas.Node][]byte` per each output that needs it. We can even have it as a helper type I guess.

We get:
1. less allocation per each sample tags we need - arguably what we wanted
2. an even more optimized tag marshalling to json or w/e actually as we can use the same thing for the csv marshal

Reply:
Actually ... we don't even use this for the creation of the node in the first place ...

So yeah it seems quite bad at thsi point. 

Arguably it's keeping some backwards compatibility ... but I am pretty sure we still breaking any usage of this type anywhere either way.

Reply:
> I would argue the json caching likely is ... uh at best overrated. 

Yeah, once this TODO is implemented, which would be super easy to do, marshaling an atlas node to JSON should be super easy: https://github.com/grafana/k6/blob/44b62ba4144079459815abc100fceb833f022532/metrics/sample.go#L245-L251

However, I suggest we can keep the `SampleTags`, for now. I am fine with getting rid of the JSON caching, but I am using them for the unindexable tags. Remember, a metric sample will have the atlas Node (for the indexed tags) and a `map[string]string` for the high-cardinality unindexed ones. It's easiest to keep `SampleTags` as a place to put both, otherwise, I will also need to change the other types like `ConnectedSampleContainer` and `ConnectedSamples` here.

Reply:
Hmmm maybe it makes more sense to have the unindexable "tags" in a separate struct either way :shrug:

Reply:
Why? :confused: It makes perfect sense for me to reuse `SampleTags`:
 - it will contain all of the "tags" for a "sample", both the indexed and the unindexed ones, I can't think of a better name
 - it will wrap the TagSet and `map[string]string` and either prevent their modification after a Sample altogether, or at least heavily discourage it

I mean, the type itself will be substantially changed - I will need to adjust some of the methods and add new ones, and also need to change how outputs work with it for unindexable tags. But I'll need a custom type and I can't think of a better name for it, so why not just reuse and change `SampleTags`? :sweat_smile: It will also narrow the scope of the things @codebien needs to fix today, since I will take care of it next week.

Reply:
We had a video call to resolve this and the notes are in https://github.com/grafana/k6/pull/2594#issuecomment-1205273413